add-interactive.c

```diff
@@ -186,3 +244,82 @@ static struct file_stat **list_modified(struct repository *r,
 	free(s);
 	return files;
 }
+
+static void populate_wi_changes(struct strbuf *buf, struct adddel *ad,
+					char *no_changes)
+{
+	if (ad->add || ad->del)
+		strbuf_addf(buf, "+%"PRIuMAX"/-%"PRIuMAX,
+			    ad->add, ad->del);
+	else
+		strbuf_addf(buf, "%s", _(no_changes));
+}
+
+static struct choices *list_and_choose(struct choices *data,

```

Comment:
I understand where Junio comes from, but C is simply not Perl. And neither is it object-oriented: we do not have a way to subclass the array items.

However, my original idea to use that `union` was to allow us to easily maintain the array as one big block that could be `free()`d in one go when it is time to clean up.

But that is not possible, anyway, because we have variable-length strings that we want to store in the array items, hence our use of `FLEXPTR_ALLOC_STR()`.

So I guess Junio's idea is a lot less bad than I initially thought, and you could pseudo-subclass `struct choice` by something like this:

```c
struct command_choice {
    struct choice choice;
    void (*command_fn)(void);
};

struct file_choice {
    struct choice choice;
    struct {
        uintmax_t added, deleted;
    } index, worktree;
};
```

Of course, you could not use `make_choice()` as-is, and in fact, I would probably not even abstract this into its own function anymore, as it would boil down to something like

```c
struct file_choice *choice;
FLEXPTR_ALLOC_STR(choice, choice.name, file_name);
```

And given that you have to declare the variable in the caller *anyway*, this indirection via `make_choice()` (which now would have to be two separate functions, `make_command_choice()` and `make_file_choice()`) would not save lines, but add lines instead.

Now, *my* opinion does not matter now. It's yours. And it is not Junio's decision to make, it is yours. Do you like his idea of keeping this so general?

Reply:
I really like his idea, but at the same time, I think this approach is much more suitable for Perl than for C. And at some point, those differences between `struct`s must be taken care of...

Reply:
Let's talk about this tomorrow. I could lend a hand...